src/domain/AnomalyDetection.ts

```diff
@@ -0,0 +1,127 @@
+import { LocationDBInterface } from "../db/location";
+import { MySQLDBInterface } from "../db/mysql";
+
+export default class AnomalyDetection {
+  locationDB: LocationDBInterface;
+  mysqlDB: MySQLDBInterface;
+  constructor(locationDB: LocationDBInterface, mysqlDB: MySQLDBInterface) {
+    this.locationDB = locationDB;
+    this.mysqlDB = mysqlDB;
+  }
+  // TODO: Add tests for this function
+  async checkForAnomaly(userId: string, newIpAddress: string) {
+    try {
+      const { latitude, longitude } = this.locationDB.getLocationBasedOnIPAddress(newIpAddress);
+      const ipAddressLog = await this.mysqlDB.getLastIPAddressLog(userId);
+
+      let riskScore = 0;
+      let status = "";
+      let message = undefined;
+      // If there is no Ip address in the database, then just update it and return
+      // TODO: remove the hard coded value of 0
+      if (ipAddressLog === undefined) {
+        await this.mysqlDB.insertLastIpAddressLog(userId, newIpAddress, latitude, longitude);
```

Comment:
how do you handle race conditions wherein you read from the db above, but then by the time the execution comes here, there is already an entry in the db for this user. In this case, the insert will fail and cause a 500 error.

Reply:
Solutions: 
1. Use `INSERT IGNORE`. It works for this specific case because there is a high chance that if race condition is happening, the request is a duplicate. 
2. Read and write can also be clubbed together in a `Transaction`. Transactions are also helpful when dealing with microservices architecture to prevent DB from getting into an inconsistent state. Isolation Level could be Read Uncommitted for this case. In use cases like reservation systems, Serializable Isolation level is better.
3. Retry is another way to handle this but not recommended here.
4. Even more granular locking mechanisms are available for Table and Rows along with different type of access locks which are not required here.